index.js

```diff
@@ -634,7 +635,12 @@ function processParams (params, layer, called, req, res, done) {
     if (!fn) return param()
 
     try {
-      fn(req, res, paramCallback, paramVal, key.name)
+      var ret = fn(req, res, paramCallback, paramVal, key.name)
+      if (isPromise(ret)) {
+        ret.then(null, function (error) {
+          paramCallback(error || new Error('Rejected promise'))
```

Comment:
```suggestion
          var err  = error || new Error('Rejected promise from param callback')
          if (Error.captureStackTrace) {
            Error.captureStackTrace(fn, err)
          }
          paramCallback(error || new Error('Rejected promise'))
```

I might be wrong, so this would be an awesome test to add, but on a rejected promise I think the error will have a short stack trace.  I *think* this would ensure that the stack has more value when reported.  Again, I think testing this would be great.  If not the stack trace I think a message better than `rejected promise` would be good.

Reply:
This was just copying the current behavior for handling rejections from middlewares, for example. Ideally what is done here is consistent between params and middlewares, so if we do think it should change, we should queue up a separate PR that can land either before or after this one with said changes.

Now, that said, this change is unlikely to do anything, as the first argument will skip the stack frames up to where that function is. So in essence, this call reduces the number of stack frames in the trace, but you seem to be saying you think there is not enough as there is?

Perhaps you can show what the before/after looks like for this change?

Reply:
I put together a sandbox, which does not test the `router.param` feature, but tests the rejected middleware feature.

https://codesandbox.io/s/github/jonchurch/express-issues/tree/async-stacks

There's three routes to test: `'/'` `'/test'` and `'/middleware'`

Please do take a look.

Reply:
Yeah I think my point was that while we are making changes, it might be nice to also improve the stack traces.  In this case, and the other middleware cases, we can provide better debugability with `.captureStackTrace`.  The example from @jonchurch seems to illustrate the issue but typically see this with normal callback code so I would have to test a bit to be sure.

The goal would be having a trace which included where the middleware function was defined, pointing to user code over express code.

Reply:
Yep, definately @wesleytodd ! I took a look at @jonchurch sandbox, and the error stack trace seems to do exactly that: it shows the user-location and not a bunch of stack within Express. Perhaps I'm just still not understanding what is missing here. Can you provide a similar sandbox showing what you are saying the issue is @wesleytodd ?

For reference, the sandbox shows the error as
```
Error: Where did I occur?
    at wait.then (/sandbox/index.js:19:37)
```

And this is like 19 +/- 1:
```js
app.get('/', (req, res, next) =>
  wait(3).then(() => Promise.reject(new Error('Where did I occur?')))
);
```

Reply:
After testing the suggested changes (on Node.js v14) I didn't find a difference between the stack traces.

Instead of testing the changes specifically to `router.param` I tested the equivalent change to `router.handleRequest` which is where the logic in this PR was copied from. 


```md
**As per PR**

/
Error: Where did I occur?
    at /home/box/Projects/express-issues/index.js:19:37

/test
Error: Where did I throw?
    at iThrow (/home/box/Projects/express-issues/anotherModule.js:2:9)
    at /home/box/Projects/express-issues/index.js:9:9

/middleware
Error: Where did I throw?
    at iThrow (/home/box/Projects/express-issues/anotherModule.js:2:9)
    at badMiddleware (/home/box/Projects/express-issues/index.js:23:9)
    at Layer.handleRequest (/home/box/Projects/router/lib/layer.js:103:15)
    at next (/home/box/Projects/router/lib/route.js:142:13)
    at Route.dispatch (/home/box/Projects/router/lib/route.js:107:3)
    at handle (/home/box/Projects/router/index.js:422:11)
    at Layer.handleRequest (/home/box/Projects/router/lib/layer.js:103:15)
    at /home/box/Projects/router/index.js:289:22
    at processParams (/home/box/Projects/router/index.js:569:12)
    at next (/home/box/Projects/router/index.js:283:5)

---------------------------

**With proposed changes:**

/
Error: Where did I occur?
    at /home/box/Projects/express-issues/index.js:19:37

/test
Error: Where did I throw?
    at iThrow (/home/box/Projects/express-issues/anotherModule.js:2:9)
    at /home/box/Projects/express-issues/index.js:9:9

/middleware
Error: Where did I throw?
    at iThrow (/home/box/Projects/express-issues/anotherModule.js:2:9)
    at badMiddleware (/home/box/Projects/express-issues/index.js:23:9)
    at Layer.handleRequest (/home/box/Projects/router/lib/layer.js:102:15)
    at next (/home/box/Projects/router/lib/route.js:142:13)
    at Route.dispatch (/home/box/Projects/router/lib/route.js:107:3)
    at handle (/home/box/Projects/router/index.js:422:11)
    at Layer.handleRequest (/home/box/Projects/router/lib/layer.js:102:15)
    at /home/box/Projects/router/index.js:289:22
    at processParams (/home/box/Projects/router/index.js:569:12)
    at next (/home/box/Projects/router/index.js:283:5)    
```